--- a/packages/internal/src/core/types.ts
+++ b/packages/internal/src/core/types.ts
@@ -230,10 +230,6 @@
 }[Issue['code']]
 
 // Impl
-interface StepMethodContext {
-	createInitialSchema: (method: string, params?: readonly unknown[]) => Use<Valchecker>
-}
-
 export interface StepMethodUtils<
 	Input,
 	Output,
@@ -290,7 +286,9 @@
 								Def[M]['Meta']['SelfIssue']
 							>
 							params: MethodTuple[0]
-							context: StepMethodContext
+							context: {
+								createInitialSchema: (method: string, params?: readonly unknown[]) => Use<Valchecker>
+							}
 						},
 					) => void
 				: never
